File: modules/extractor.py
"""Information extraction module for resumes (email, phone, education, skills)."""

import re
from pathlib import Path
import spacy
from spacy.matcher import PhraseMatcher
import config

# Initialize lightweight English blank model (no model download needed)
_nlp = spacy.blank("en")
_matcher = PhraseMatcher(_nlp.vocab, attr="LOWER")


def _load_skills_matcher():
    """Load skill phrases from skills.txt into the spaCy PhraseMatcher once at import time."""
    skills_file = Path(config.SKILLS_TXT)
    if not skills_file.exists():
        return

    with open(skills_file, "r", encoding="utf-8") as f:
        skill_phrases = [line.strip().lower() for line in f if line.strip()]

    # Convert phrase strings into spaCy Doc objects for PhraseMatcher
    patterns = [_nlp.make_doc(phrase) for phrase in skill_phrases]
    _matcher.add("SKILLS", patterns)


# Build skills matcher once at import time
_load_skills_matcher()

# Regular expressions for contact information
_EMAIL_REGEX = re.compile(r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+")
_PHONE_REGEX = re.compile(
    r"(?:(?:\+?\d{1,3}[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4})|(?:\+?\d{1,3}[-.\s]?)?\b\d{10}\b|(?:\+\d{1,3}[-.\s]?\d{5}[-.\s]?\d{5}))"
)

# Education degree keywords specified in SRS FR-2
_DEGREE_KEYWORDS = [
    r"\bb\.?tech\b",
    r"\bbca\b",
    r"\bmca\b",
    r"\bb\.?sc\b",
    r"\bm\.?sc\b",
    r"\bmba\b",
    r"\bbachelor\b",
    r"\bmaster\b",
    r"\bphd\b",
    r"\bdiploma\b",
    r"\bb\.?e\.?\b",
    r"\bm\.?tech\b",
]
_DEGREE_PATTERN = re.compile("|".join(_DEGREE_KEYWORDS), re.IGNORECASE)


def extract_info(text: str) -> dict:
    """Extract email, phone, education, and skills from raw resume text.

    Args:
        text: Raw text content of the resume.

    Returns:
        dict: A dictionary containing:
            - email: str or None
            - phone: str or None
            - education: list[str] (max 5 matching lines)
            - skills: list[str] (unique, lowercase, sorted)
    """
    # Extract email
    email_match = _EMAIL_REGEX.search(text)
    email = email_match.group(0) if email_match else None

    # Extract phone
    phone_match = _PHONE_REGEX.search(text)
    phone = phone_match.group(0).strip() if phone_match else None

    # Extract education lines (matching degree keywords, maximum 5 lines)
    education_lines = []
    for line in text.splitlines():
        cleaned_line = line.strip()
        if cleaned_line and _DEGREE_PATTERN.search(cleaned_line):
            # Ignore false matches like 'Scrum Master' unless an actual degree keyword is also present
            if re.search(r"\bscrum master\b", cleaned_line, re.IGNORECASE) and not re.search(
                r"\b(bachelor|degree|university|college|institute|graduat|b\.?tech|m\.?tech|bca|mca|b\.?sc|m\.?sc|mba|phd|diploma)\b",
                cleaned_line,
                re.IGNORECASE,
            ):
                continue
            if cleaned_line not in education_lines:
                education_lines.append(cleaned_line)
            if len(education_lines) >= 5:
                break

    # Extract skills using spaCy PhraseMatcher
    doc = _nlp(text)
    matches = _matcher(doc)
    extracted_skills = set()

    for match_id, start, end in matches:
        span = doc[start:end]
        extracted_skills.add(span.text.lower())

    # Return unique, lowercase, alphabetically sorted skills
    skills = sorted(list(extracted_skills))

    return {
        "email": email,
        "phone": phone,
        "education": education_lines,
        "skills": skills,
    }

File: src/extractor.ts
import fs from 'fs';
import { SKILLS_TXT } from './config.js';

export interface CandidateInfo {
  email: string | null;
  phone: string | null;
  education: string[];
  skills: string[];
}

// Regular expressions for contact information matching Python specifications
const EMAIL_REGEX = /[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+/;
const PHONE_REGEX = /(?:(?:\+?\d{1,3}[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4})|(?:\+?\d{1,3}[-.\s]?)?\b\d{10}\b|(?:\+\d{1,3}[-.\s]?\d{5}[-.\s]?\d{5}))/;

const DEGREE_KEYWORDS = [
  /\bb\.?tech\b/i,
  /\bbca\b/i,
  /\bmca\b/i,
  /\bb\.?sc\b/i,
  /\bm\.?sc\b/i,
  /\bmba\b/i,
  /\bbachelor\b/i,
  /\bmaster\b/i,
  /\bphd\b/i,
  /\bdiploma\b/i,
  /\bb\.?e\.?\b/i,
  /\bm\.?tech\b/i,
];

const DEGREE_VALIDATION = /\b(bachelor|degree|university|college|institute|graduat|b\.?tech|m\.?tech|bca|mca|b\.?sc|m\.?sc|mba|phd|diploma)\b/i;

// Preload skills from skills.txt
let skillPhrases: string[] = [];
interface SkillMatcher {
  skill: string;
  regex: RegExp;
}
let skillMatchers: SkillMatcher[] = [];

function escapeRegExp(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function initSkills(skillsFilePath: string = SKILLS_TXT) {
  if (!fs.existsSync(skillsFilePath)) {
    return;
  }
  const fileContent = fs.readFileSync(skillsFilePath, 'utf-8');
  skillPhrases = fileContent
    .split(/\r?\n/)
    .map((s) => s.trim().toLowerCase())
    .filter(Boolean);

  skillMatchers = skillPhrases.map((skill) => {
    const escaped = escapeRegExp(skill);
    const startBound = /^[a-zA-Z0-9]/.test(skill) ? '(?<=^|[^a-zA-Z0-9])' : '';
    const endBound = /[a-zA-Z0-9]$/.test(skill) ? '(?=[^a-zA-Z0-9]|$)' : '';
    return {
      skill,
      regex: new RegExp(`${startBound}${escaped}${endBound}`, 'i'),
    };
  });
}

// Initialize skills at module load
initSkills();

/**
 * Extract email, phone, education, and skills from raw resume text.
 */
export function extractInfo(text: string): CandidateInfo {
  if (!text || typeof text !== 'string') {
    return {
      email: null,
      phone: null,
      education: [],
      skills: [],
    };
  }

  // 1. Email extraction
  const emailMatch = text.match(EMAIL_REGEX);
  const email = emailMatch ? emailMatch[0] : null;

  // 2. Phone extraction
  const phoneMatch = text.match(PHONE_REGEX);
  const phone = phoneMatch ? phoneMatch[0].trim() : null;

  // 3. Education extraction
  const educationLines: string[] = [];
  const lines = text.split(/\r?\n/);
  for (const rawLine of lines) {
    const cleanedLine = rawLine.trim();
    if (!cleanedLine) continue;

    const matchesDegree = DEGREE_KEYWORDS.some((kw) => kw.test(cleanedLine));
    if (matchesDegree) {
      // Disqualify 'scrum master' unless an actual degree word is also present
      if (/\bscrum master\b/i.test(cleanedLine) && !DEGREE_VALIDATION.test(cleanedLine)) {
        continue;
      }

      if (!educationLines.includes(cleanedLine)) {
        educationLines.push(cleanedLine);
      }
      if (educationLines.length >= 5) {
        break;
      }
    }
  }

  // 4. Skills extraction
  const extractedSkills = new Set<string>();
  for (const { skill, regex } of skillMatchers) {
    if (skill === 'c') {
      // Avoid matching 'c' in general sentences unless explicitly referencing C language
      if (/(?:programming\s+in\s+c|\bc\s*[,/]\s*c\+\+|\bc\s*[,/]|skills:.*\bc\b)/i.test(text)) {
        extractedSkills.add('c');
      }
      continue;
    }
    if (regex.test(text)) {
      extractedSkills.add(skill);
    }
  }

  const skills = Array.from(extractedSkills).sort();

  return {
    email,
    phone,
    education: educationLines,
    skills,
  };
}
